services/optimization.js

```diff
@@ -0,0 +1,162 @@
+var stringifyBooleans = require('./service-helpers/stringify-booleans');
+
+/**
+ * Optimization API service.
+ *
+ * Learn more about this service and its responses in
+ * [the HTTP service documentation](https://www.mapbox.com/api-documentation/#optimization).
+ */
+var Optimization = {};
+
+/**
+ * Get a duration-optimized route.
+ *
+ * Please read [the full HTTP service documentation](https://www.mapbox.com/api-documentation/#optimization)
+ * to understand all of the available options.
+ *
+ * @param {Object} config
+ * @param {'driving'|'walking'|'cycling'} [config.profile="driving"]
+ * @param {Array<OptimizationWaypoint>} config.waypoints - An ordered array of [`OptimizationWaypoint`](#optimizationwaypoint) objects, between 2 and 12 (inclusive).
+ * @param {Array<'duration'|'distance'|'speed'>} [config.annotations] - Specify additional metadata that should be returned.
+ * @param {'any' | 'last'} [config.destination="any"] Returned route ends at `any` or `last` coordinate.
+ * @param {[number, number]} [config.distributions] Array of objects, each of which includes a `pickup` and `dropoff` property. `pickup` and `dropoff` properties correspond to an index in the coordinates array.
+ * @param {'geojson'|'polyline'|'polyline6'} [config.geometries="polyline"] - Format of the returned geometry.
+ * @param {string} [config.language="en"] - Language of returned turn-by-turn text instructions.
+ *   See options listed in [the HTTP service documentation](https://www.mapbox.com/api-documentation/#instructions-languages).
+ * @param {'simplified'|'full'|'false'} [config.overview="simplified"] - Type of returned overview geometry.
+ * @param {boolean} [config.roundtrip=true] Specifies whether the trip should complete by returning to the first location. 
+ * @param {'any' | 'first'} [config.source="any"] To begin the route, start either from the first coordinate or let the Optimization API choose.
+ * @param {boolean} [config.steps=false] - Whether to return steps and turn-by-turn instructions.
+ 
+ * @return {MapiRequest}
+ */
+Optimization.getOptimization = function(config) {
+  v.assertShape({
+    profile: v.oneOf('driving', 'walking', 'cycling'),
+    waypoints: v.required(
+      v.arrayOf(
+        v.shape({
+          coordinates: v.required(v.coordinates),
+          approach: v.oneOf('unrestricted', 'curb'),
+          bearing: v.arrayOf(v.range([0, 360])),
+          radius: v.oneOfType(v.number, v.equal('unlimited'))
+        })
+      )
+    ),
+    annotations: v.arrayOf(v.oneOf('duration', 'distance', 'speed')),
+    geometries: v.string,
+    language: v.string,
+    overview: v.oneOf('simplified', 'full', 'false'),
+    roundtrip: v.boolean,
+    steps: v.boolean,
+    source: v.oneOf('any', 'first'),
+    destination: v.oneOf('any', 'last'),
+    distributions: v.arrayOf(
+      v.shape({
+        pickup: v.number,
+        dropoff: v.number
+      })
+    )
+  })(config);
+
+  config.profile = config.profile || 'driving';
```

Comment:
If we've been using this pattern in other places (which we probably have), we should change it. This will mutate the object that the user passed in — which we don't want to do.

Reply:
Moved this check down to `createRequest` instead:

```
params: {
      profile: config.profile || 'driving',
      coordinates: path.coordinates.join(';')
    }
```